filters/auth/grant.go

```diff
@@ -0,0 +1,239 @@
+}
+
+func (f grantFilter) refreshTokenIfRequired(c cookie) (*oauth2.Token, error) {
+	now := time.Now()
+	isAccessTokenExpired := c.isAccessTokenExpired()
+	canRefresh := !f.config.DisableRefresh && c.RefreshToken != ""
+	shouldRefresh := isAccessTokenExpired || now.After(c.RefreshAfter)
+
+	var token *oauth2.Token
+	var err error
+
+	if canRefresh && shouldRefresh {
+		log.Debugf(
+			"Refreshing token. Can refresh: %v, Should refresh: %v, Expired: %v",
+			canRefresh, shouldRefresh, isAccessTokenExpired,
+		)
+
+		token, err = f.refreshToken(c)
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return token, err
+}
+
+func (f grantFilter) checkAccessTokenValidity(ctx filters.FilterContext, c cookie) (map[string]interface{}, error) {
+	if c.isAccessTokenExpired() {
+		return nil, errExpiredToken
+	}
+
+	tokeninfo, err := f.config.TokeninfoClient.getTokeninfo(c.AccessToken, ctx)
+	if err != nil {
+		return nil, err
+	}
+	return tokeninfo, nil
+}
+
+func (f grantFilter) setAccessTokenHeader(req *http.Request, token string) {
+	if f.config.AccessTokenHeaderName != "" {
+		req.Header.Set(f.config.AccessTokenHeaderName, authHeaderPrefix+token)
+	}
+}
+
+func (f grantFilter) createTokenContainer(token *oauth2.Token, tokeninfo map[string]interface{}) tokenContainer {
+	subject := ""
+	if f.config.TokeninfoSubjectKey != "" {
+		subject = tokeninfo[f.config.TokeninfoSubjectKey].(string)
+	}
+
+	tokeninfo["sub"] = subject
+
+	return tokenContainer{
+		OAuth2Token: token,
+		Subject:     subject,
+		Claims:      tokeninfo,
+	}
+}
+
+func (f grantFilter) Request(ctx filters.FilterContext) {
+	req := ctx.Request()
+
+	c, err := getCookie(req, f.config)
+	if err == http.ErrNoCookie {
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	tokeninfo, err := f.checkAccessTokenValidity(ctx, *c)
+	if err != nil && err != errExpiredToken {
+		if err != errInvalidToken {
+			log.Errorf("Error while calling tokeninfo: %v.", err)
+		}
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	token, err := f.refreshTokenIfRequired(*c)
+	if err != nil && c.isAccessTokenExpired() {
+		// Refresh failed and we no longer have a valid access token.
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	if token == nil {
+		// Token can be null if:
+		// 1. No refresh was necessary, so it is not yet initialized.
+		// 2. Refresh failed, but we still have a valid access token.
+		//    Let the request processing continue and defer the login
+		//    redirect until the access token becomes invalid.
+		token = &oauth2.Token{
+			AccessToken:  c.AccessToken,
+			TokenType:    "Bearer",
+			RefreshToken: c.RefreshToken,
+			Expiry:       c.Expiry,
+		}
+	}
+
+	if tokeninfo == nil {
+		// We need to fetch tokeninfo again if we refreshed the token.
+		tokeninfo, err = f.config.TokeninfoClient.getTokeninfo(token.AccessToken, ctx)
+		if err != nil {
+			if err == errInvalidToken {
+				log.Errorf("Error while calling tokeninfo: %v.", err)
+			}
+			loginRedirect(ctx, f.config)
+			return
+		}
+	}
```

Comment:
This codepath feels complicated. If `checkAccessTokenValidity` is inlined it becomes apparent that there is `f.config.TokeninfoClient.getTokeninfo` twice in the code path. Maybe if `refreshTokenIfRequired` is changed to make token from cookie when refresh is not needed like:
```go
func (f grantFilter) refreshTokenIfRequired(c cookie) (*oauth2.Token, error) {
	now := time.Now()
	isAccessTokenExpired := c.isAccessTokenExpired()
	canRefresh := !f.config.DisableRefresh && c.RefreshToken != ""
	shouldRefresh := isAccessTokenExpired || now.After(c.RefreshAfter)

	if canRefresh && shouldRefresh {
		log.Debugf(
			"Refreshing token. Can refresh: %v, Should refresh: %v, Expired: %v",
			canRefresh, shouldRefresh, isAccessTokenExpired,
		)

		return f.refreshToken(c)
	}
	return &oauth2.Token{
		AccessToken:  c.AccessToken,
		TokenType:    "Bearer",
		RefreshToken: c.RefreshToken,
		Expiry:       c.Expiry,
	}, nil
}
```
this codepath could be simplified to
```go
        token, err := f.refreshTokenIfRequired(*c)
	if err != nil {
		// Refresh failed and we no longer have a valid access token.
		loginRedirect(ctx, f.config)
		return
	}

	tokeninfo, err := f.config.TokeninfoClient.getTokeninfo(token.AccessToken, ctx)
	if err != nil {
		if err == errInvalidToken {
			log.Errorf("Error while calling tokeninfo: %v.", err)
		}
		loginRedirect(ctx, f.config)
		return
	}
```
WDYT?

Reply:
Yes makes sense. I modified it to what you suggested with two differences:

1. Right after refreshing token, I keep the second condition `c.isAccessTokenExpired()` here:
```go
	token, err := f.refreshTokenIfRequired(*c)
	if err != nil && c.isAccessTokenExpired() {
```
The reason is that if the access token is not yet expired, but the refresh call failed (because the refresh is done well before the expiry), then we don't want to force the user to relogin at this point since they still have a valid access token (or if it is revoked, that gets caught by the later tokeninfo call).

2. Right after doing the tokeninfo check, I added a condition to check that the returned `tokeninfo` is not nil
```go
	tokeninfo, err := f.config.TokeninfoClient.getTokeninfo(token.AccessToken, ctx)
	if err != nil || tokeninfo == nil {
```
To be honest, I can't think of a scenario where it would happen, but if it does, it will break downstream filters unexpectedly.

Reply:
Nice!

Regarding 1. `if err != nil && c.isAccessTokenExpired() ` I can not easily get it. 
> The reason is that if the access token is not yet expired, but the refresh call failed 

Ok, that means we are refreshing because of `now.After(c.RefreshAfter)`. Is that `c.RefreshAfter` not reliable?

Regarding 2. `if err != nil || tokeninfo == nil` I would leave just error check and rather fix tokeninfoclient to provide strict contract "either error or valid result" if we ever observe nil result (I don't see how that is possible).

IMO we should strive for strict method contracts in general such that it is either error or valid return value to have simple error checking `if err != nil`; otherwise it is hard to reason and reader stumbles and has to parse error checking conditions.

Reply:
> Ok, that means we are refreshing because of now.After(c.RefreshAfter). Is that c.RefreshAfter not reliable?

As you saw the `RefreshAfter` is the expiry duration / 10. So if the token is set to expire in 1h, it will actually try the first refresh after 6 mins. I don't know exactly why `/ 10` was chosen, but if the refresh call fails then a user's session will be invalidated a lot before the actual token expiry.

> Regarding 2. if err != nil || tokeninfo == nil I would leave just error check

I reviewed the tokeninfoclient code and there is no way tokeninfo can be null. Also, an error is returned in all cases. So I agree the `tokeninfo == nil` is not needed. I will remove it.

Reply:
2. Fixed